File: python-implementation/src/dynamic_discovery/agent_b_server.py
"""
Agent B - Math Calculator Agent
===============================

Questo agente implementa il protocollo Agent2Agent (A2A) per calcoli matematici avanzati.
Fornisce capacità di calcolo matematico attraverso JSON-RPC 2.0 con supporto real-time.

Funzionalità matematiche:
- Operazioni base: add, subtract, multiply, divide
- Operazioni avanzate: power, sqrt, factorial
- Funzioni trigonometriche: sin, cos, tan (in gradi)
- Gestione errori matematici (divisione per zero, radici negative, etc.)
- Validazione input e controllo tipi

Endpoints conformi A2A:
- GET /.well-known/agent.json - Agent Card per discovery
- GET /status - Health check con statistiche
- POST /rpc - JSON-RPC 2.0 per calcoli
- GET /events - Server-Sent Events per aggiornamenti
- GET /api/tasks - Monitoring task matematici

Porta: 3002
Dominio: mathematics
Formati: application/json
"""

from flask import Flask, request, jsonify, Response
from flask_cors import CORS
import json
import time
import threading
import uuid
import math
from datetime import datetime

# Inizializzazione Flask con CORS per cross-origin requests
app = Flask(__name__)
CORS(app)  # Abilita CORS per compatibilità multi-dominio

# Configurazione agente matematico con metadati specifici
AGENT_CONFIG = {
    "id": "agent-b-math-calculator",          # ID unico per discovery
    "name": "Math Calculator Agent",          # Nome descrittivo
    "description": "Advanced mathematical calculations and operations",
    "version": "2.0.0",                       # Versione agente
    "port": 3002,                            # Porta di ascolto
    # Capacità matematiche esposte per discovery automatico
    "capabilities": [
        "math.basic.add",                     # Addizione
        "math.basic.subtract",                # Sottrazione  
        "math.basic.multiply",                # Moltiplicazione
        "math.basic.divide",                  # Divisione
        "math.advanced.power",                # Elevamento a potenza
        "math.advanced.sqrt",                 # Radice quadrata
        "math.advanced.factorial",            # Fattoriale
        "math.trigonometry.sin",              # Seno (gradi)
        "math.trigonometry.cos",              # Coseno (gradi)
        "math.trigonometry.tan"               # Tangente (gradi)
    ]
}

# Storage per task matematici attivi - tracking asincrono
active_tasks = {}

# Storage per aggiornamenti task - utilizzato per Server-Sent Events
task_updates = {}

# Agent Card specializzata per dominio matematico
AGENT_CARD = {
    "agent": {
        "id": AGENT_CONFIG["id"],
        "name": AGENT_CONFIG["name"],
        "description": AGENT_CONFIG["description"],
        "version": AGENT_CONFIG["version"],
        "capabilities": AGENT_CONFIG["capabilities"],
        "endpoints": {
            "rpc": f"http://localhost:{AGENT_CONFIG['port']}/rpc",
            "status": f"http://localhost:{AGENT_CONFIG['port']}/status",
            "events": f"http://localhost:{AGENT_CONFIG['port']}/events"
        },
        "discovery": {
            "registrationUrl": "http://localhost:3010/api/agents/register",
            "lastRegistered": None
        },
        "metadata": {
            "type": "math-calculator",                          # Tipo specializzato
            "domain": "mathematics",                            # Dominio matematico
            "supportedFormats": ["application/json"],           # Solo JSON
            "authMethods": ["none", "api-key"],
            "rateLimit": {"requests": 100, "window": 60}
        }
    },
    "spec": {
        "protocol": "agent2agent",
        "version": "1.0.0",
        "schema": "https://schemas.agent2agent.ai/v1/agent-card.json"
    }
}

def process_math_task(task_id, method, params):
    """
    Processa task di calcolo matematico in modo asincrono.
    
    Gestisce tutte le operazioni matematiche supportate dall'agente:
    - Operazioni base: add, subtract, multiply, divide
    - Operazioni avanzate: power, sqrt, factorial  
    - Funzioni trigonometriche: sin, cos, tan
    
    Caratteristiche:
    - Validazione rigorosa input (controllo tipi, divisione per zero)
    - Gestione errori matematici specifici
    - Conversione gradi->radianti per trigonometria
    - Tracking progressi real-time per SSE
    
    Args:
        task_id (str): Identificativo unico del task matematico
        method (str): Metodo JSON-RPC chiamato  
        params (dict): Parametri contenenti 'operation' e 'numbers'
        
    Parametri supportati:
        operation (str): Tipo operazione matematica
        numbers (list): Lista numeri per calcolo
        
    Note:
        - Risultati includono timestamp e input originali
        - Errori matematici gestiti con messaggi specifici
        - Aggiornamenti real-time via task_updates per SSE
    """
    try:
        # Inizializza tracking aggiornamenti per questo task
        task_updates[task_id] = []
        
        def add_update(status, message, data=None):
            """
            Aggiunge aggiornamento di stato per tracking real-time.
            
            Args:
                status (str): Stato operazione ('processing', 'completed', 'error')
                message (str): Messaggio descrittivo progresso
                data (dict, optional): Dati aggiuntivi (risultati calcolo)
            """
            update = {
                "timestamp": datetime.utcnow().isoformat(),
                "status": status,
                "message": message,
                "data": data
            }
            task_updates[task_id].append(update)
        
        add_update("processing", "Starting mathematical calculation")
        
        # Estrazione parametri con default sicuri
        operation = params.get("operation", "add")
        numbers = params.get("numbers", [])
        
        # Validazione input - numeri obbligatori per calcoli
        if not numbers:
            add_update("error", "No numbers provided")
            active_tasks[task_id] = {"status": "error", "error": "No numbers provided"}
            return
        
        add_update("processing", f"Performing operation: {operation}")
        
        # Engine di calcolo matematico con gestione errori specializzata
        try:
            # === OPERAZIONI ARITMETICHE BASE ===
            if operation == "add":
                # Somma di tutti i numeri nella lista
                result = sum(numbers)
                
            elif operation == "subtract":
                # Sottrazione sequenziale: primo - secondo - terzo...
                result = numbers[0]
                for num in numbers[1:]:
                    result -= num
                    
            elif operation == "multiply":
                # Moltiplicazione di tutti i numeri
                result = 1
                for num in numbers:
                    result *= num
                    
            elif operation == "divide":
                # Divisione sequenziale con controllo divisione per zero
                result = numbers[0]
                for num in numbers[1:]:
                    if num == 0:
                        raise ValueError("Division by zero")
                    result /= num
                    
            # === OPERAZIONI AVANZATE ===
            elif operation == "power":
                # Elevamento a potenza: base^esponente
                if len(numbers) < 2:
                    raise ValueError("Power operation requires base and exponent")
                result = numbers[0] ** numbers[1]
                
            elif operation == "sqrt":
                # Radice quadrata con controllo numeri negativi
                if len(numbers) != 1:
                    raise ValueError("Square root requires exactly one number")
                if numbers[0] < 0:
                    raise ValueError("Cannot calculate square root of negative number")
                result = math.sqrt(numbers[0])
                
            elif operation == "factorial":
                # Fattoriale con controllo intero non negativo
                if len(numbers) != 1:
                    raise ValueError("Factorial requires exactly one number")
                if numbers[0] < 0 or not isinstance(numbers[0], int):
                    raise ValueError("Factorial requires a non-negative integer")
                result = math.factorial(int(numbers[0]))
                
            # === FUNZIONI TRIGONOMETRICHE (gradi -> radianti) ===
            elif operation == "sin":
                # Seno con conversione automatica gradi->radianti
                if len(numbers) != 1:
                    raise ValueError("Sine requires exactly one number")
                result = math.sin(math.radians(numbers[0]))
                
            elif operation == "cos":
                # Coseno con conversione automatica gradi->radianti  
                if len(numbers) != 1:
                    raise ValueError("Cosine requires exactly one number")
                result = math.cos(math.radians(numbers[0]))
                
            elif operation == "tan":
                # Tangente con conversione automatica gradi->radianti
                if len(numbers) != 1:
                    raise ValueError("Tangent requires exactly one number")
                result = math.tan(math.radians(numbers[0]))
                
            else:
                # Operazione non supportata
                add_update("error", f"Unknown operation: {operation}")
                active_tasks[task_id] = {"status": "error", "error": f"Unknown operation: {operation}"}
                return
              # Costruzione risultato strutturato con metadati
            calculation_result = {
                "operation": operation,                        # Operazione eseguita
                "inputs": numbers,                            # Input originali
                "result": result,                             # Risultato calcolo
                "timestamp": datetime.utcnow().isoformat()    # Timestamp esecuzione
            }
            
            # Task completato con successo
            add_update("processing", "Mathematical calculation completed")
            add_update("completed", "Task completed successfully", calculation_result)
            
            # Salva risultato finale nel storage task attivi
            active_tasks[task_id] = {
                "status": "completed",
                "result": calculation_result,
                "completedAt": datetime.utcnow().isoformat()
            }
            
        except Exception as calc_error:
            # Gestione errori di calcolo specifici (matematici)
            add_update("error", f"Calculation failed: {str(calc_error)}")
            active_tasks[task_id] = {"status": "error", "error": str(calc_error)}
        
    except Exception as e:
        # Gestione errori generali di processamento
        add_update("error", f"Processing failed: {str(e)}")
        active_tasks[task_id] = {"status": "error", "error": str(e)}

@app.route('/.well-known/agent.json')
def agent_card():
    """Serve the Agent Card"""
    return jsonify(AGENT_CARD)

@app.route('/status')
def status():
    """Health check endpoint"""
    return jsonify({
        "status": "ok",
        "agent": AGENT_CONFIG["name"],
        "version": AGENT_CONFIG["version"],
        "timestamp": datetime.utcnow().isoformat(),
        "activeTasks": len(active_tasks)
    })

@app.route('/rpc', methods=['POST'])
def handle_rpc():
    """Handle JSON-RPC 2.0 requests"""
    try:
        data = request.get_json()
        
        if not data or data.get('jsonrpc') != '2.0':
            return jsonify({
                "jsonrpc": "2.0",
                "error": {"code": -32600, "message": "Invalid Request"},
                "id": data.get('id') if data else None
            }), 400
        
        method = data.get('method')
        params = data.get('params', {})
        request_id = data.get('id')
        
        if method == 'agent.getCapabilities':
            return jsonify({
                "jsonrpc": "2.0",
                "result": {
                    "capabilities": AGENT_CONFIG["capabilities"],
                    "agent": AGENT_CONFIG["name"],
                    "version": AGENT_CONFIG["version"]
                },
                "id": request_id
            })
        
        elif method == 'tasks.send':
            task_id = str(uuid.uuid4())
            
            # Start processing in background
            thread = threading.Thread(
                target=process_math_task,
                args=(task_id, method, params)
            )
            thread.start()
            
            return jsonify({
                "jsonrpc": "2.0",
                "result": {
                    "taskId": task_id,
                    "status": "accepted",
                    "message": "Math task accepted for processing"
                },
                "id": request_id
            })
        
        elif method == 'tasks.status':
            task_id = params.get('taskId')
            if task_id in active_tasks:
                return jsonify({
                    "jsonrpc": "2.0",
                    "result": active_tasks[task_id],
                    "id": request_id
                })
            else:
                return jsonify({
                    "jsonrpc": "2.0",
                    "error": {"code": -32602, "message": "Task not found"},
                    "id": request_id
                }), 404
        
        else:
            return jsonify({
                "jsonrpc": "2.0",
                "error": {"code": -32601, "message": "Method not found"},
                "id": request_id
            }), 404
            
    except Exception as e:
        return jsonify({
            "jsonrpc": "2.0",
            "error": {"code": -32603, "message": f"Internal error: {str(e)}"},
            "id": data.get('id') if 'data' in locals() else None
        }), 500

@app.route('/events')
def events():
    """Server-Sent Events endpoint for real-time updates"""
    def generate():
        yield "data: " + json.dumps({
            "type": "connected",
            "agent": AGENT_CONFIG["name"],
            "timestamp": datetime.utcnow().isoformat()
        }) + "\n\n"
        
        # Send task updates
        sent_updates = set()
        while True:
            for task_id, updates in task_updates.items():
                for i, update in enumerate(updates):
                    update_key = f"{task_id}_{i}"
                    if update_key not in sent_updates:
                        yield "data: " + json.dumps({
                            "type": "task_update",
                            "taskId": task_id,
                            "update": update
                        }) + "\n\n"
                        sent_updates.add(update_key)
            
            time.sleep(1)
    
    return Response(generate(), mimetype='text/event-stream')

@app.route('/api/tasks', methods=['GET'])
def get_tasks():
    """Get all active tasks"""
    return jsonify({
        "tasks": active_tasks,
        "count": len(active_tasks)
    })

if __name__ == '__main__':
    print(f"🚀 Starting {AGENT_CONFIG['name']} v{AGENT_CONFIG['version']}")
    print(f"📡 Agent Card: http://localhost:{AGENT_CONFIG['port']}/.well-known/agent.json")
    print(f"🔍 Status: http://localhost:{AGENT_CONFIG['port']}/status")
    print(f"⚡ RPC Endpoint: http://localhost:{AGENT_CONFIG['port']}/rpc")
    print(f"📊 Events: http://localhost:{AGENT_CONFIG['port']}/events")
    
    app.run(host='0.0.0.0', port=AGENT_CONFIG['port'], debug=True)

File: src/dynamic-discovery/agent-b-server.js
import express from 'express';
import cors from 'cors';
import { v4 as uuidv4 } from 'uuid';

/**
 * Agent B - Math Calculator
 * Esegue operazioni matematiche avanzate e calcoli scientifici
 * Implementa il protocollo A2A con discovery dinamico
 */
class AgentBServer {
  constructor(port = 4002) {
    this.port = port;
    this.agentId = 'math-calculator-agent-b';
    this.app = express();
    this.tasks = new Map();
    
    this.setupMiddleware();
    this.setupRoutes();
  }

  setupMiddleware() {
    this.app.use(cors());
    this.app.use(express.json());
  }

  setupRoutes() {
    // Status endpoint per A2A
    this.app.get('/status', (req, res) => {
      res.json({
        status: 'ok',
        timestamp: new Date().toISOString(),
        agent_id: this.agentId,
        port: this.port
      });
    });

    // Agent Card per discovery dinamico
    this.app.get('/agent-card', (req, res) => {
      res.json({
        id: this.agentId,
        name: 'Agent B - Math Calculator',
        description: 'Advanced mathematical operations and scientific calculations agent',
        version: '1.0.0',
        type: 'math-calculator',
        endpoints: {
          status: '/status',
          rpc: '/rpc',
          task_status: '/task/{task_id}',
          agent_card: '/agent-card'
        },
        capabilities: [
          {
            id: 'basic-math',
            name: 'Basic Mathematics',
            description: 'Basic operations: addition, subtraction, multiplication, division',
            input_format: 'application/json',
            output_format: 'application/json'
          },
          {
            id: 'advanced-math',
            name: 'Advanced Mathematics',
            description: 'Advanced operations: power, square root, logarithm, trigonometry',
            input_format: 'application/json',
            output_format: 'application/json'
          },
          {
            id: 'statistical-analysis',
            name: 'Statistical Analysis',
            description: 'Statistical operations on number arrays',
            input_format: 'application/json',
            output_format: 'application/json'
          },
          {
            id: 'equation-solving',
            name: 'Equation Solving',
            description: 'Solves simple linear and quadratic equations',
            input_format: 'application/json',
            output_format: 'application/json'
          }
        ],
        authentication: {
          type: 'none'
        },
        discovery_info: {
          discoverable: true,
          category: 'mathematics',
          tags: ['calculator', 'math', 'statistics', 'equations']
        }
      });
    });

    // JSON-RPC 2.0 endpoint
    this.app.post('/rpc', async (req, res) => {
      try {
        const { jsonrpc, method, params, id } = req.body;

        // Validazione JSON-RPC 2.0
        if (jsonrpc !== '2.0') {
          return res.status(400).json({
            jsonrpc: '2.0',
            error: { code: -32600, message: 'Invalid Request' },
            id
          });
        }

        // Gestione dei metodi supportati
        const supportedMethods = ['basic_math', 'advanced_math', 'statistical_analysis', 'equation_solving'];
        
        if (!supportedMethods.includes(method)) {
          return res.status(400).json({
            jsonrpc: '2.0',
            error: { code: -32601, message: `Method not found: ${method}` },
            id
          });
        }

        const taskId = uuidv4();
        
        // Crea task
        this.tasks.set(taskId, {
          id: taskId,
          status: 'processing',
          method,
          params,
          created_at: new Date().toISOString(),
          agent: 'agent-b'
        });

        // Elabora task in modo asincrono
        this.processTask(taskId, method, params);

        // Risposta immediata A2A
        res.json({
          jsonrpc: '2.0',
          result: { 
            task_id: taskId, 
            status: 'accepted',
            message: `Math task ${method} accepted for processing`
          },
          id
        });

      } catch (error) {
        res.status(500).json({
          jsonrpc: '2.0',
          error: { code: -32603, message: 'Internal error', data: error.message },
          id: req.body.id
        });
      }
    });

    // Task status endpoint
    this.app.get('/task/:taskId', (req, res) => {
      const task = this.tasks.get(req.params.taskId);
      if (!task) {
        return res.status(404).json({ error: 'Task not found' });
      }
      res.json(task);
    });
  }

  // Elaborazione dei task
  async processTask(taskId, method, params) {
    const task = this.tasks.get(taskId);
    if (!task) return;

    try {
      // Simula elaborazione matematica
      await new Promise(resolve => setTimeout(resolve, 800));

      let result;

      switch (method) {
        case 'basic_math':
          result = this.performBasicMath(params);
          break;
        case 'advanced_math':
          result = this.performAdvancedMath(params);
          break;
        case 'statistical_analysis':
          result = this.performStatisticalAnalysis(params);
          break;
        case 'equation_solving':
          result = this.solveEquation(params);
          break;
        default:
          throw new Error(`Unsupported method: ${method}`);
      }

      // Aggiorna task con risultato
      task.status = 'completed';
      task.result = result;
      task.completed_at = new Date().toISOString();
      
      console.log(`🔵 Agent B: Task ${taskId} completed - ${method}`);

    } catch (error) {
      task.status = 'failed';
      task.error = error.message;
      task.failed_at = new Date().toISOString();
      
      console.error(`❌ Agent B: Task ${taskId} failed:`, error.message);
    }
  }

  // Operazioni matematiche base
  performBasicMath(params) {
    const { operation, a, b } = params;

    if (typeof a !== 'number' || (operation !== 'sqrt' && typeof b !== 'number')) {
      throw new Error('Invalid numeric parameters');
    }

    let result;
    let formula;

    switch (operation) {
      case 'add':
        result = a + b;
        formula = `${a} + ${b} = ${result}`;
        break;
      case 'subtract':
        result = a - b;
        formula = `${a} - ${b} = ${result}`;
        break;
      case 'multiply':
        result = a * b;
        formula = `${a} × ${b} = ${result}`;
        break;
      case 'divide':
        if (b === 0) throw new Error('Division by zero');
        result = a / b;
        formula = `${a} ÷ ${b} = ${result}`;
        break;
      case 'modulo':
        if (b === 0) throw new Error('Modulo by zero');
        result = a % b;
        formula = `${a} mod ${b} = ${result}`;
        break;
      default:
        throw new Error(`Unsupported basic operation: ${operation}`);
    }

    return {
      operation: operation,
      operands: operation === 'sqrt' ? [a] : [a, b],
      result: result,
      formula: formula,
      type: 'basic_math',
      processed_by: 'agent-b',
      timestamp: new Date().toISOString()
    };
  }

  // Operazioni matematiche avanzate
  performAdvancedMath(params) {
    const { operation, a, b } = params;

    if (typeof a !== 'number') {
      throw new Error('Invalid numeric parameter');
    }

    let result;
    let formula;

    switch (operation) {
      case 'power':
        if (typeof b !== 'number') throw new Error('Power requires two parameters');
        result = Math.pow(a, b);
        formula = `${a}^${b} = ${result}`;
        break;
      case 'sqrt':
        if (a < 0) throw new Error('Square root of negative number');
        result = Math.sqrt(a);
        formula = `√${a} = ${result}`;
        break;
      case 'cbrt':
        result = Math.cbrt(a);
        formula = `∛${a} = ${result}`;
        break;
      case 'log':
        if (a <= 0) throw new Error('Logarithm of non-positive number');
        result = Math.log(a);
        formula = `ln(${a}) = ${result}`;
        break;
      case 'log10':
        if (a <= 0) throw new Error('Logarithm of non-positive number');
        result = Math.log10(a);
        formula = `log₁₀(${a}) = ${result}`;
        break;
      case 'sin':
        result = Math.sin(a);
        formula = `sin(${a}) = ${result}`;
        break;
      case 'cos':
        result = Math.cos(a);
        formula = `cos(${a}) = ${result}`;
        break;
      case 'tan':
        result = Math.tan(a);
        formula = `tan(${a}) = ${result}`;
        break;
      case 'factorial':
        if (a < 0 || !Number.isInteger(a)) throw new Error('Factorial requires non-negative integer');
        result = this.calculateFactorial(a);
        formula = `${a}! = ${result}`;
        break;
      default:
        throw new Error(`Unsupported advanced operation: ${operation}`);
    }

    return {
      operation: operation,
      operands: ['power'].includes(operation) ? [a, b] : [a],
      result: result,
      formula: formula,
      type: 'advanced_math',
      processed_by: 'agent-b',
      timestamp: new Date().toISOString()
    };
  }

  // Analisi statistica
  performStatisticalAnalysis(params) {
    const { numbers } = params;

    if (!Array.isArray(numbers) || numbers.length === 0) {
      throw new Error('Numbers array is required and cannot be empty');
    }

    if (!numbers.every(n => typeof n === 'number')) {
      throw new Error('All elements must be numbers');
    }

    const sorted = [...numbers].sort((a, b) => a - b);
    const sum = numbers.reduce((acc, n) => acc + n, 0);
    const mean = sum / numbers.length;
    const variance = numbers.reduce((acc, n) => acc + Math.pow(n - mean, 2), 0) / numbers.length;
    const standardDeviation = Math.sqrt(variance);
    
    // Mediana
    const median = sorted.length % 2 === 0 
      ? (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2
      : sorted[Math.floor(sorted.length / 2)];

    // Moda
    const frequency = {};
    numbers.forEach(n => frequency[n] = (frequency[n] || 0) + 1);
    const maxFreq = Math.max(...Object.values(frequency));
    const mode = Object.keys(frequency)
      .filter(n => frequency[n] === maxFreq)
      .map(n => parseFloat(n));

    return {
      data: numbers,
      statistics: {
        count: numbers.length,
        sum: sum,
        mean: mean,
        median: median,
        mode: mode.length === numbers.length ? 'No mode' : mode,
        variance: variance,
        standard_deviation: standardDeviation,
        min: Math.min(...numbers),
        max: Math.max(...numbers),
        range: Math.max(...numbers) - Math.min(...numbers)
      },
      type: 'statistical_analysis',
      processed_by: 'agent-b',
      timestamp: new Date().toISOString()
    };
  }

  // Risoluzione equazioni
  solveEquation(params) {
    const { type, coefficients } = params;

    if (!type || !Array.isArray(coefficients)) {
      throw new Error('Equation type and coefficients array are required');
    }

    let solution;
    let formula;

    switch (type) {
      case 'linear':
        // ax + b = 0
        if (coefficients.length !== 2) {
          throw new Error('Linear equation requires 2 coefficients [a, b]');
        }
        const [a, b] = coefficients;
        if (a === 0) throw new Error('Coefficient a cannot be zero for linear equation');
        solution = { x: -b / a };
        formula = `${a}x + ${b} = 0 → x = ${solution.x}`;
        break;

      case 'quadratic':
        // ax² + bx + c = 0
        if (coefficients.length !== 3) {
          throw new Error('Quadratic equation requires 3 coefficients [a, b, c]');
        }
        const [qa, qb, qc] = coefficients;
        if (qa === 0) throw new Error('Coefficient a cannot be zero for quadratic equation');
        
        const discriminant = qb * qb - 4 * qa * qc;
        
        if (discriminant < 0) {
          solution = { complex: true, message: 'No real solutions (complex roots)' };
        } else if (discriminant === 0) {
          const x = -qb / (2 * qa);
          solution = { x1: x, x2: x, note: 'One solution (repeated root)' };
        } else {
          const x1 = (-qb + Math.sqrt(discriminant)) / (2 * qa);
          const x2 = (-qb - Math.sqrt(discriminant)) / (2 * qa);
          solution = { x1: x1, x2: x2 };
        }
        
        formula = `${qa}x² + ${qb}x + ${qc} = 0`;
        break;

      default:
        throw new Error(`Unsupported equation type: ${type}`);
    }

    return {
      equation_type: type,
      coefficients: coefficients,
      formula: formula,
      solution: solution,
      type: 'equation_solving',
      processed_by: 'agent-b',
      timestamp: new Date().toISOString()
    };
  }

  // Funzione ausiliaria per fattoriale
  calculateFactorial(n) {
    if (n === 0 || n === 1) return 1;
    let result = 1;
    for (let i = 2; i <= n; i++) {
      result *= i;
    }
    return result;
  }

  start() {
    this.app.listen(this.port, () => {
      console.log(`🔵 Agent B (Math Calculator) running on http://localhost:${this.port}`);
      console.log(`📋 Agent Card: http://localhost:${this.port}/agent-card`);
      console.log(`✅ Ready for A2A discovery and communication`);
    });
  }
}

// Avvia server se eseguito direttamente
if (process.argv[1] && process.argv[1].endsWith('agent-b-server.js')) {
  const agentB = new AgentBServer();
  agentB.start();
}

export default AgentBServer;
